src/radical/pilot/agent/bootstrap_0.sh

```diff
@@ -993,19 +1001,81 @@ virtenv_create()
 
     virtenv="$1"
     python_dist="$2"
+    virtenv_dist="$3"
 
     if test "$python_dist" = "default"
     then
 
-        VIRTENV_CMD="$PYTHON -m venv"
+        # check if the `venv` module is usable
+        if run_cmd "Create ve with venv" \
+                   "python3 -m venv $virtenv"
+        then
+            true
 
-        run_cmd "Create virtualenv" \
-                "$VIRTENV_CMD $virtenv"
+        else
+            # the `venv` module is not available - fall back to virtualenv
+            if test "$virtenv_dist" = "default"
+            then
+                virtenv_dist="20.21.0"
+            fi
 
-        if test $? -ne 0
+            if test "$virtenv_dist" = "20.21.0"
+            then
+                flags='-1 -k -L -O'
+                if (hostname -f | grep -e '^smic' > /dev/null)
+                then
+                    flags='-k -L -O'
+                fi
+
+                run_cmd "Download virtualenv tgz" \
+                        "curl $flags '$VIRTENV_TGZ_URL'"
+
+                if ! test "$?" = 0
+                then
+                    echo "WARNING: couldn't download virtualenv via curl! Using system version"
+                    virtenv_dist="system"
+
+                else
+                    run_cmd "unpacking virtualenv tgz" \
+                            "tar zxmf '$VIRTENV_TGZ'"
+
+                    if test $? -ne 0
+                    then
+                        echo "Couldn't unpack virtualenv! Using system version"
+                        virtenv_dist="system"
+                    else
+                        VIRTENV_CMD="$PYTHON $VIRTENV_DIR/virtualenv.py"
+                    fi
+
+                fi
+            fi
+
+            # don't use `elif` here - above falls back to 'system' virtenv on errors
+            if test "$virtenv_dist" = "system"
+            then
+                VIRTENV_CMD="virtualenv"
+            fi
+
+            if test "$VIRTENV_CMD" = ""
+            then
+                echo "ERROR: invalid or unusable virtenv_dist option"
+                return 1
+            fi
+
+            run_cmd "Create ve with virtualenv" \
+                    "$VIRTENV_CMD $virtenv"
+
+            if test $? -ne 0
+            then
+                echo "ERROR: couldn't create virtualenv"
+                return 1
+            fi
+        fi
+
+        # clean out virtenv sources
+        if test -d "$VIRTENV_DIR"
         then
-            echo "ERROR: couldn't create virtualenv"
-            return 1
```

Comment:
should we just install `virtualenv` with `pip`, if `venv` is unavailable?
```
pip install virtualenv
VIRTENV_CMD="virtualenv"
```

also I don't think we need to return back `VIRTENV_DIST`, we can just check (1) `venv`, (2) `virtualenv` through `pip`, (3) a direct link for `virtualenv-16.7.5` (but I don't think we need this third option)

p.s. `"$virtenv_dist" = "20.21.0"` - this versioning is unclear, since it is not related to `virtualenv`, but to `curl` flags; and that bringing back to my point about `virtualenv_dist`

Reply:
check `venv`, if not check `virtualenv`,  if not download `virtualenv`.